openeogeotrellis/identity.py

```diff
@@ -0,0 +1,124 @@
+    public_key: str
+    """The public key in PEM format which can be used to verify the tokens"""
+
+    @classmethod
+    def from_file(cls, file_path: Path) -> "IDPDetails":
+        with open(file_path) as fh:
+            file_dict = json.load(fh)
+        private_key = file_dict["private_key"].strip()
+        cls.assert_valid_pem(private_key, "private")
+        public_key = file_dict["public_key"].strip()
+        cls.assert_valid_pem(public_key, "public")
+        return cls(
+            file_dict["issuer"],
+            private_key,
+            public_key
+        )
+
+    @staticmethod
+    def assert_valid_pem(key: str, key_type: Literal["private", "public"]):
+        """Check if a string follows the expected structure of a PEM formatted key"""
+        template = '-----{location} RSA {key_type} KEY-----'
+        pem_header = template.format(location="BEGIN", key_type=key_type.upper())
+        pem_footer = template.format(location="END", key_type=key_type.upper())
+        assert key.startswith(pem_header), f"Key did not start with {pem_header}"
+        assert key.endswith(pem_footer), f"Key did not end with {pem_footer}"
+
+
+class IDPTokenIssuer:
+    """
+    An issuer of signed tokens. This is a singleton so to avoid errors having duplicate instances you can call the
+    class method `instance() to get the singleton object`.
+    """
+    _IDP_DETAILS_FILE = Path(os.environ.get("OPENEO_IDP_DETAILS_FILE", "/opt/.idp/idp_details.json"))
+    _TOKEN_EXPIRES_IN_SECONDS = 48 * 3600
+    _SINGLETON: Optional[IDPTokenIssuer] = None
+
+    def __init__(self):
+        if self.__class__._SINGLETON is not None:
+            raise RuntimeError("IDPTokenIssuer is meant as a singleton")
+        self._IDP_DETAILS: Optional[IDPDetails] = None
+        self.__class__._SINGLETON = self
+
+    @classmethod
+    def instance(cls) -> "IDPTokenIssuer":
+        if cls._SINGLETON is not None:
+            return cls._SINGLETON
+        else:
+            return cls()
+
+    def _reload_idp_details_if_needed(self) -> None:
+        register_reload = get_file_reload_register_func_if_changed(self._IDP_DETAILS_FILE)
+        if register_reload is None:
+            return
+        try:
+            self._IDP_DETAILS = IDPDetails.from_file(self._IDP_DETAILS_FILE)
+            register_reload()
+        except AssertionError as ae:
+            logger.fatal(f"Invalid config staged for idp: {ae}")
+        except Exception as e:
+            logger.warning(f"Could not reload IDP details from {self._IDP_DETAILS_FILE} due to {e}")
+
+    def get_identity_token(self, user_id: str, job_id: str) -> Optional[str]:
+        self._reload_idp_details_if_needed()
+        if self._IDP_DETAILS is None:
+            return None
+        else:
+            idp_details = self._IDP_DETAILS
+            now = utcnow()
+            return jwt.encode(
+                {
+                    "sub": user_id,  # TODO: see if we can use subject from upstream IDP instead
```

Comment:
@soxofaan I wanted to check this with you before having to dive to deep in the code. In a way it is OK to have user_id because that is what we use to identify users in web driver and during execution. But since the user_id is also a claim in the token it could also make sense to keep the subject of the initial token (so the openeo access token) but I am not sure whether that is easily retrievable at this stage.

Reply:
I guess you actually need the "sub" identifier here, which is the same as user_id in most cases I guess, but I'm not sure we can guarantee that (I also think we have a ticket somewhere to actually decouple the user_id from the "sub" identifier). So I think it's best to be explicit here and work with a variable "sub_id" here instead of "user_id".   

Another reason to avoid "user_id" here are client credential flows where the sub id is associated with the client, and not the user

Reply:
I can modify this function to take sub_id but I don't know how/where to get the sub identifier in this openeo-geopyspark-driver module (particularly during the start job method in backend.py).

Reply:
yes, I would make that new function future-proof and generic with a honest argument name "sub_id".

And where you call it you can then make it more explicit that we are doing a bit of user_id->sub_id mapping, which is fine for now:
```
.get_identity_token(sub_id=user_id, job_id=job_id)
```